File: core/workflow/nodes/requirement_category.py
import sys, uuid, json
from pathlib import Path
sys.path.append(str(Path(__file__).absolute().parent.parent.parent.parent))

from typing import Dict, List, Optional, Union, Any
from datetime import datetime
from . import LLMBaseNode
from ..variables import Variable, ArrayVariable, ObjectVariable, validate_required_variable, get_first_variable_value
from ..context import Context, replace_variable_value_with_context
from core.llm.prompt import Prompt
from languages import get_language_content
from log import Logger


logger = Logger.get_logger('celery-app')

class RequirementCategory:
    """
    A class to manage requirement categories, each represented as a dictionary with an ID and content.
    """
    
    def __init__(self):
        """
        Initializes the RequirementCategory object with an empty list to store categories.
        """
        self.categories: List[Dict[str, str]] = []

    def add_category(self, content: str) -> None:
        """
        Adds a new category to the list of categories.

        :param content: The content of the requirement category.
        """
        category = {
            "id": str(uuid.uuid4()),
            "content": content
        }
        self.categories.append(category)
        
    def get_category_name_by_id(self, category_id: str) -> str:
        """
        Retrieves the name of the category given its ID.

        :param category_id: The ID of the category to find.
        :return: The content (name) of the category if found, otherwise returns an empty string.
        """
        for category in self.categories:
            if category["id"] == category_id:
                return category["content"]
        return ""
        
    def to_dict(self) -> List[Dict[str, str]]:
        """
        Converts the list of categories to a list of dictionaries.

        :return: A list of dictionaries, each representing a requirement category.
        """
        return self.categories
    
def create_requirement_category_from_dict(categories_list: List[Dict[str, str]]) -> RequirementCategory:
    """
    Creates an instance of RequirementCategory from a list of dictionaries.

    :param categories_list: A list of dictionaries, each representing a requirement category.
    :return: An instance of RequirementCategory populated with the provided categories.
    """
    requirement_category = RequirementCategory()
    requirement_category.categories = categories_list
    return requirement_category

class RequirementCategoryNode(LLMBaseNode):
    """
    A RequirementCategoryNode object is used to create requirement categories in a workflow.
    """
    
    def __init__(
        self, 
        title: str, 
        desc: str = "", 
        input: Optional[Union[Variable, ArrayVariable, ObjectVariable]] = None, 
        model_config_id: int = 0, 
        requirement_category: Optional[RequirementCategory] = None, 
        prompt: Optional[Prompt] = None,
        flow_data: Dict[str, Any] = {}, 
        original_node_id: Optional[str] = None
    ):
        """
        Initializes a RequirementCategoryNode object with typing enhancements and the ability to track the original node ID.
        """
        init_kwargs = {
            "type": "requirement_category",
            "title": title,
            "desc": desc,
            "input": input,
            "model_config_id": model_config_id,
            "requirement_category": requirement_category,
            "prompt": prompt,
            "flow_data": flow_data
        }
        if original_node_id is not None:
            init_kwargs["original_node_id"] = original_node_id
        
        super().__init__(**init_kwargs)
        
    def run(
        self,
        context: Context,
        app_run_id: int = 0,
        edge_id: str = '',
        **kwargs
    ) -> Dict[str, Any]:
        """
        Execute the node task.
        """
        try:
            start_time = datetime.now()
            requirement_category = self.data['requirement_category']
            if not requirement_category:
                raise Exception("Requirement category not found.")
            category_example_json = json.dumps({'id': 'category ID', 'content': 'category name'})
            all_category_json = json.dumps(requirement_category.to_dict(), ensure_ascii=False)
            
            input = self.data['input']
            replace_variable_value_with_context(input, context)
            validate_required_variable(input)
            
            prompt_config = get_language_content("requirement_category")
            if self.data["prompt"]:
                prompt_config["system"] += "\n" + self.duplicate_braces(self.data["prompt"].get_system())
            self.data["prompt"] = Prompt(system=prompt_config["system"], user=prompt_config["user"])
            
            model_data, category, prompt_tokens, completion_tokens, total_tokens = self.invoke(
                app_run_id=app_run_id, 
                edge_id=edge_id,
                context=context, 
                input={
                    'format': category_example_json,
                    'requirement': get_first_variable_value(input),
                    'categories': all_category_json
                },
                return_json=True,
            )
            
            category_id = category["id"]
            category_name = requirement_category.get_category_name_by_id(category_id)
            if not category_name:
                raise Exception("Category name not found.")
            outputs = Variable(name="category_name", type="string", value=category_name)
            
            end_time = datetime.now()
            return {
                'status': 'success',
                'message': 'Requirement category node executed successfully.',
                'data': {
                    'elapsed_time': end_time.timestamp() - start_time.timestamp(),
                    'inputs': input.to_dict(),
                    'model_data': model_data,
                    'condition_id': category_id,
                    'output_type' : 1,
                    'outputs': outputs.to_dict(),
                    'prompt_tokens': prompt_tokens,
                    'completion_tokens': completion_tokens,
                    'total_tokens': total_tokens
                }
            }
        except Exception as e:
            logger.exception('ERROR!!')
            return {
                'status': 'failed',
                'message': str(e)
            }

File: web/src/py2js/nodes/requirement_category.js
// Importing necessary modules and classes
import { v4 as uuidv4 } from 'uuid';
import { Prompt } from '../prompt.js';
import { VariableTypes } from '../variables.js';
import { Node } from './base.js';

/**
 * A class to manage requirement categories, each represented as an object with an ID and content.
 */
export class RequirementCategory {
    constructor() {
        /**
         * Initializes the RequirementCategory object with an empty array to store categories.
         */
        this.categories = [];
    }

    /**
     * Adds a new category to the array of categories.
     *
     * @param {string} content The content of the requirement category.
     */
    addCategory(content, id) {
        const category = {
            id: id || uuidv4(),
            content: content,
        };
        this.categories.push(category);
    }

    /**
     * Retrieves the name of the category given its ID.
     *
     * @param {string} categoryId The ID of the category to find.
     * @return {string} The content (name) of the category if found, otherwise returns an empty string.
     */
    getCategoryNameById(categoryId) {
        const category = this.categories.find(cat => cat.id === categoryId);
        return category ? category.content : '';
    }

    /**
     * Converts the array of categories to an array of objects.
     *
     * @return {Object[]} An array of objects, each representing a requirement category.
     */
    toObject() {
        return this.categories;
    }
}

/**
 * Creates an instance of RequirementCategory from an array of objects.
 *
 * @param {Object[]} categoriesList An array of objects, each representing a requirement category.
 * @return {RequirementCategory} An instance of RequirementCategory populated with the provided categories.
 */
export function createRequirementCategoryFromObject(categoriesList) {
    const requirementCategory = new RequirementCategory();
    requirementCategory.categories = categoriesList;
    return requirementCategory;
}

/**
 * A RequirementCategoryNode object is used to create requirement categories in a workflow.
 * Inherits from the Node class.
 */
export class RequirementCategoryNode extends Node {
    /**
 * Initialize a RequirementCategoryNode object with specific attribute requirements.
 * The backend output defaults to a string type Variable named 'category_name'.
 *
 * @param {Object} params - An object containing initialization parameters.
 * @param {string} params.title - The title of the node.
 * @param {string} [params.desc=""] - The description of the node.
 * @param {VariableTypes|null} [params.input=null] - The input variables for the node.
 * @param {number} [params.model_config_id=0] - The model configuration ID.
 * @param {RequirementCategory|null} [params.requirement_category=null] - The requirement category associated with the node.
 * @param {Prompt|null} [params.prompt=null] - The prompt associated with the node.
 * @param {boolean} [params.manual_confirmation=false] - Whether manual confirmation is required.
 * @param {Object} [params.position={}] - The position of the node.
 * @param {number} [params.width=0] - The width of the node.
 * @param {number} [params.height=0] - The height of the node.
 * @param {boolean} [params.selected=false] - Whether the node is selected.
 * @param {string|null} [params.original_node_id=null] - The original node ID, if any.
 */
    constructor({
        title,
        desc = '',
        input = null,
        model_config_id = 0,
        requirement_category = null,
        prompt = null,
        manual_confirmation = false,
        position = {},
        flow_data = {},
        width = 0,
        height = 0,
        selected = false,
        original_node_id = null,
    }) {
        // Constructing the initialization object
        const initKwargs = {
            type: 'requirement_category',
            title,
            desc,
            input,
            model_config_id,
            requirement_category,
            prompt,

            flow_data,
        };

        // Adding the original node ID if provided
        if (original_node_id !== null) {
            initKwargs.original_node_id = original_node_id;
        }

        // Calling the parent class constructor
        super(initKwargs);
    }
}

// Exporting the RequirementCategoryNode class for use in other modules
